### DISCUSSION TITLE
Offline sync

### DISCUSSION THREAD

**Discussion by heyhippari** [jellyfin/jellyfin-vue#808] [Features]
The recent Kakao/Spotify debacle motivated me enough to start thinking about how to do this. This is very early and meanly meant for brainstorming, but it's always good to have an idea of where to start when we get there.

So, in keeping with the store structure in #801, we could imagine an added `availableOffline` array with the index of items we cached.

`canPlay` in the `itemHelper` would be adjusted to look at `$nuxt.isOnline` and/or `$nuxt.isOffline`.

If we are online, it would proceed as usual. However, if we are offline, it would look at the `availableOffline` array to see if the item is cached, then mark it as playable if it is cached.

For the caching process itself, we can use [`cache.add()`](https://developer.mozilla.org/en-US/docs/Web/API/Cache/add) and [workbox](https://developers.google.com/web/tools/workbox/guides/advanced-recipes#cached-av).

Shaka Player also [supports offline playback natively](https://github.com/google/shaka-player/blob/master/docs/tutorials/offline.md).

There's the whole thing of being able to choose tracks and bandwidth too, but that would likely hook up to the normal playback system.

**Comment by ThibaultNocchi** [comment]
I think we should prioritize offline playback even when online. Why waste bandwidth and kill baby dolphins when you can use your cached version?

I also think we could have cache and synced stuff, like Spotify. Cache would just keep played stuff and be cleared from a single button. Synced would be manually marked and downloaded stuff, which you could review somewhere and wouldn't get cleaned by the clear cache button.

Wdyt?

**Reply by heyhippari** [reply]
The normal cache would be tracked by the browser itself. We have a bit less control than an app here, since both that and the sync would use the same cache.

Though we can use [this](https://developer.mozilla.org/en-US/docs/Web/API/Cache/delete) to manually clear entries, so it could be done for both.

I also agree we should be favoring the cached thing if it is, which would be done automatically, unless the selected tracks are different, so that should be mostly taken care of for us :yum:

**Reply by ferferga** [reply]
This is actually a really good point because in current web, for instance, if you are repeating one song, the client downloads the file over and over. I think this is already covered in Vue as we just scrobble back to 0, but honestly I have no idea what happens if we go forward one song and then back.

As MrTim mentions, Shaka handles automatically most of the stuff + the exposed APIs for more granular control if we ever need it, but I never checked this myself manually and it's a really good point.

**Comment by ferferga** [comment]
This is not 100% on topic, but I think we should accomplish this before we even start with offline stuff.

``isOnline`` and ``isOffline`` is great when there's zero connectivity. As soon as the device has an IP it will appear as connected to the browser, but not if there's actual internet connection. We should investigate ways to "extend" it or create our own to handle the following cases:

• DNS resolution issues.
• Server is offline (this will probably cover also DNS issues)
• Websocket connection lost (this alone will probably cover all the cases above)

Right now Websocket can also recover from a prolonged disconnection. Best way to reproduce it is to leave the client open, put your computer in standby mode for a few hours and come back. Lots of exceptions will be thrown and basically you must relogin.

**Reply by ThibaultNocchi** [reply]
Shouldn't websockets be a pretty good way to say we are connected to the server?

**Reply by ferferga** [reply]
Yep, forgot to say that we can cover with it server offline problem (the same way I did with DNS resolution) :P

Edited

**Reply by heyhippari** [reply]
This is what Nuxt uses behind the scenes, I believe: https://developer.mozilla.org/en-US/docs/Web/API/NavigatorOnLine/onLine

Imo, that's as far as we should go to detect if we're offline or not. That + websocket state should be fairly telling.

**Reply by ThibaultNocchi** [reply]
IMO we should just check for connection to the server

**Reply by heyhippari** [reply]
We don't want to run requests constantly to check that, though. Relying on the browser telling us if we're online is the best course of action, imo.

That and the websocket being online covers 99.9% of use cases.

**Reply by ThibaultNocchi** [reply]
Yeah ofc, I was more thinking about checking the WS state. It should pretty reliable, shouldn't it?

**Reply by heyhippari** [reply]
Sort of. The websocket itself isn't linked to the connection status.

Of course, it'll drop if the connection drops, and it retries a few times, but we could be into a state, sometimes, where the connection is up but the websocket is down.

Even though that shouldn't happen, it introduced a failure point which we might not want.

**Reply by ThibaultNocchi** [reply]
What are those situations? And if the websockets are down, doesn't it mean that the connection isn't stable and streaming music would be hazardous?

**Comment by deadrabbit87** [comment]
Hi, 

maybe helpful: I found the android client Finamp which is able to sync music offline. 

https://github.com/UnicornsOnLSD/finamp

On my phone i works like a charm, much better and more stable than the original Jellyfin-App. 

BW

**Reply by GlassedSilver** [reply]
Does it sync back plays when offline once online again?

**Reply by ThibaultNocchi** [reply]
I don't think, you can ask the creator on the Matrix chan he's usually around
